Modify EqualityList and FlowsToList into EqualityMap and FlowsToMap accordingly

feliciahalim opened the pull request:
By modify `EqualityList` and `FlowsToList`  into map, we can retrieve the element faster by using find with that would take O(log n) compare to O(n) in vector/list. 

There's no subsumption difference in `klee-examples/basic` and `scalability/Regexp`.
When I run `scalability/Regexp` with option` -max-fail-subsume=3`, this PR reduces the execution time from 43.23 to 24.43

For `make-test` run, one more test case has passed and execution time is also faster.
Before:
********************
Testing Time: 616.14s
********************
Failing Tests (4):
    KLEE :: Feature/MemoryLimit.c
    KLEE :: Runtime/POSIX/DirConsistency.c
    KLEE :: Runtime/POSIX/DirSeek.c
    KLEE :: Runtime/POSIX/Write2.c

  Expected Passes    : 164
  Expected Failures  : 2
  Unsupported Tests  : 2
  Unexpected Failures: 4
After:
```
********************
Testing Time: 607.48s
********************
Failing Tests (3):
    KLEE :: Feature/MemoryLimit.c
    KLEE :: Runtime/POSIX/DirConsistency.c
    KLEE :: Runtime/POSIX/DirSeek.c

  Expected Passes    : 165
  Expected Failures  : 2
  Unsupported Tests  : 2
  Unexpected Failures: 3

```

feliciahalim commented:
I have just modified `valuesList` into `valuesMap`, the result is the same as stated in the previous comment, with faster execution time for scalability/Regexp with option-max-fail-subsume=3 from 24.43 to 20.80

domainexpert commented:
@feliciahalim Some comments:
1. I suggest the implementation of addDependency to call addDependencyViaAllocation so as not to replicate code.
2. Please use 0 instead of NULL (0 is more of C++ standard).
3. Please remove the declaration and implementation of FlowsTo class.

feliciahalim commented:
@domainexpert Thank you for your suggestion. I have updated the code accordingly.

domainexpert commented:
@feliciahalim I prefer Line 1530 of Dependency.cpp to be `stream << "=={";` and Line 1540 to be `stream << "}]";`. This may make the output easier to read by distinguishing the pointer at the left-hand side, and the allocations it is equated with.

domainexpert commented:
@feliciahalim There was a segmentation fault when I uncommented Lines 2866-2872 of Executor.cpp.

domainexpert commented:
@feliciahalim The segmentation fault is on `addition.c` example of `basic` directory of https://github.com/tracer-x/klee-examples.

feliciahalim commented:
@domainexpert Thank you for noticing the segmentation fault. It is caused by an attempt to print a null pointer as when we store flow dependency via allocation, allocation can be null. I have fixed the it by null checking before printing. I have also replaced [ to { syntax for easier to read as suggested, as well as in the flow dependency. 
Based on https://github.com/tracer-x/klee/issues/130, I have renamed the name accordingly.

domainexpert commented:
@feliciahalim Please note that in general the resolution for separate issues such as #130 better be under separate PR (to keep each PR focused), but it is ok for now as the change is small and unlikely to generate further discussions.

More importantly, there's an infinite loop when I tried to run `make addition.klee` of the `basic` directory of tracer-x/klee-examples, with Lines 2866-2872 of `Executor.cpp` uncommented.

feliciahalim commented:
@domainexpert sorry about that. I have fixed the infinite loop caused by wrong increment of iterator in Flow Dependency. Okay noted, for the next separate issue would be developed in separate PR.

domainexpert commented:
@feliciahalim Not a problem re fix for #130. For the flow dependency, in your current implementation a flow dependency element is printed as:
```
[target,{source,allocation}]
```
Perhaps:
```
[target <- source] via allocation 
```
is a better format. Also, when the dependency is not via a store-load through an allocation, (that is, allocation is 0 / NULL), I think it is enough to print
```
[target <- source]
```
without the `via` clause.

feliciahalim commented:
@domainexpert The current implementation is actually : 
`[target1, {source1, allocation1}, {source2, allocation2}, .. ]
`

In this case, is it supposed to be transformed like this:
`[target1 <- source1] via allocation1, [target1 <- source2] via allocation2 
`
Please let me know.. Thanks

domainexpert commented:
@feliciahalim Yes, I suggest so.

feliciahalim commented:
@domainexpert I have updated the printing format of Flows Dependency as suggested. Thank you.

domainexpert commented:
@feliciahalim May be good to enclose `<-` and `via` with whitespaces, i.e., at Line 1575 of Dependency.cpp, to use " <- " instead of the current "<-", and also " via " instead of "via" at Line 1579.

domainexpert commented:
@feliciahalim Awesome work! Thank you!
